### ISSUE TITLE
reactions do not work over props when used in constructor, but they work when used in componentDidMount

### ISSUE THREAD

**Issue by xaviergonz** [mobxjs/mobx-react#478]
React 16.3, latest mobx-react

Today I found a weird issue when I was creating a reaction over props in the component constructor (using TS btw):
```
@observer
class MyC extends React.Component {
  constructor(props) {
    super(props);
    this.r = reaction(() => this.props.open, () => {
      // for some reason this would only run once, but never again when the this.props.open changed
    }, { fireImmediately: true });
  }
}
```

however when I moved the reaction to componentDidMount it all started working...

makes me wonder, maybe react is internally making a copy of the props and swapping them before componentDidMount executes?

or maybe props made observable after the constructor runs?
if so then maybe the decorator should override the super class constructor function and inject the make observable thing on it?

**Comment by xaviergonz** [comment]
ps: it also works on componentWillMount, but since it is going to be deprecated that's why I started putting them in the constructor

**Comment by mweststrate** [comment]
Use didMount, as recommended by the react docs. Pretty sure it will work
😊. Until mounting happened mobx isn't tracking the props causing your
reaction not to fire

Op za 19 mei 2018 23:52 schreef Javier Gonzalez <notifications@github.com>:

**Comment by xaviergonz** [comment]
This is what I mean
https://codesandbox.io/s/6zkk0k8rv3

when I run that and I start hitting the button the console shows this:

reaction on constructor 
reaction on componentDidMount 
button clicked 
reaction on componentDidMount 
button clicked 
reaction on componentDidMount 
button clicked 
reaction on componentDidMount 
button clicked 
reaction on componentDidMount

**Comment by xaviergonz** [comment]
I mean, I'm happy with creating the reactions on componentDidMount, but it would be nice to know why and/or see that on the docs somewhere :)

**Comment by mweststrate** [comment]
PRs to the docs are welcome :)

**Comment by xaviergonz** [comment]
ah, I was just hopeful it could be fixed so it works on constructors :(

here's the PR for the docs https://github.com/mobxjs/mobx-react/pull/480

**Comment by Strate** [comment]
Same here, this issue introduced with update from 5.0.0 to 5.2.0

Props now instrumented after componentWillMount fires

**Comment by Strate** [comment]
@mweststrate I think we can make componentWillMount receive reactive props. In `observer` decorator, check that `componentWillMount` exists, if so: patch props & state there, else: patch props & state in render

**Comment by Strate** [comment]
Or, we can even try to define state/props getter&setter in observer decorator, so have reactive props/state even in constructor

**Comment by mweststrate** [comment]
I think componentWillMount detection isn't a real solution, it will just
delay the inevitable a few months. Using the observer to return a new
construction could possible work indeed, that is an interesting approach to
try

Op do 7 jun. 2018 om 08:46 schreef Artur Eshenbrener <
notifications@github.com>:

**Comment by Strate** [comment]
@mweststrate i would like to investigate on this.

**Comment by mweststrate** [comment]
Feel free! (and sorry for not really diving into the other PR yet, first
want to stabilize around React 16.3 and MobX 5. P.S: we could make that one
a MobX 5 supported version ony, and then use proxies observable objects
that might make things simpler? Not sure)

Op do 7 jun. 2018 om 10:40 schreef Artur Eshenbrener <
notifications@github.com>:

**Comment by Strate** [comment]
5.0.0 - constructor has non-observable props, componentWillMount has observable props
5.2.0 - constructor has non-observable props, componentWillMount has non-observable props

**Comment by mweststrate** [comment]
(just for background, the change in 5.2.0 is in response to React 16.3
where componentWillMount was deprecated, so willMount shouldn't be used
anymore in the first place if you are upgrading deps)

Op do 7 jun. 2018 om 15:11 schreef Artur Eshenbrener <
notifications@github.com>:

**Comment by Strate** [comment]
@mweststrate I made a PR. It was easy :)

**Comment by mweststrate** [comment]
That looks awesome!

Op do 7 jun. 2018 om 16:43 schreef Artur Eshenbrener <
notifications@github.com>:

**Comment by mweststrate** [comment]
Fixed in 5.2.1, both in constructor and willMount reactions can safely be set up again now. Thanks @Strate!
